backends/llamacpp/src/backend.rs

```diff
@@ -0,0 +1,496 @@
+
+    let ctx = unsafe { &mut *ctx };
+
+    // Append the new token to the generated ones
+    ctx.generation.generated_tokens.push(new_token_id);
+
+    // Generate response
+    let response = match ctx.tokenizer.decode(&[new_token_id], false) {
+        Ok(text) => {
+            let special = ctx.tokenizer.get_added_vocabulary().is_special_token(&text);
+            let token = Token {
+                id: new_token_id,
+                text,
+                logprob: new_token_logit,
+                special,
+            };
+
+            // Should we generate an ending or intermediate response?
+            match is_final {
+                false => Ok(InferStreamResponse::Intermediate {
+                    token,
+                    top_tokens: vec![],
+                }),
+                true => {
+                    // Decode the whole text
+                    match ctx
+                        .tokenizer
+                        .decode(&ctx.generation.generated_tokens, false)
+                    {
+                        Ok(text) => Ok(InferStreamResponse::End {
+                            token,
+                            top_tokens: vec![],
+                            generated_text: GeneratedText {
+                                text,
+                                generated_tokens: n_generated_tokens as u32,
+                                finish_reason: FinishReason::Length,
+                                seed: Some(ctx.generation.sampling_params.seed),
+                            },
+                            start: ctx.start,
+                            queued: ctx.start,
+                        }),
+                        Err(err) => Err(GenerationError(err.to_string())),
+                    }
+                }
+            }
+        }
+        Err(ref err) => Err(GenerationError(err.to_string())),
+    };
+
+    // Send back to the client
+    let status = ctx.stream.send(response).inspect_err(|err| {
+        error!("Failed to send back the response: {}", err);
+    });
+    status.is_err()
+}
+
+/// Main loop allowing scheduling incoming requests without blocking the main execution thread
+///
+/// # Arguments
+///
+/// * `queue`: Synchronized container to receive new request
+/// * `backlog`: Synchronized container to dispatch new request towards all the workers for one to pick it up.
+///
+/// returns: ()
+///
+/// # Examples
+///
+/// ```
+///
+/// ```
+async fn scheduler_loop(
+    mut queue: UnboundedReceiver<(GenerationContext, UnboundedSender<InferResult>)>,
+    backlog: MpmcSender<(GenerationContext, UnboundedSender<InferResult>)>,
+) {
+    // Let's receive incoming requests
+    loop {
+        match queue.recv().await {
+            None => break,
+            Some((ctx, sender)) => {
+                if let Err(e) = backlog.send((ctx, sender)).await {
+                    todo!("What do we do")
```

Comment:
Looks like it can only happen id the channel is closed (all receivers are drop), so I guess you can break the loop and just return / exit